infrastructure/crypto/src/keys.rs

```diff
@@ -72,4 +72,7 @@ pub trait PublicKey:
         let pk = Self::from_secret_key(&k);
         (k, pk)
     }
+
+    /// Generate a shared secret from one party's private key and another party's public key
+    fn shared_secret(k: &Self::K, pk: &Self) -> Self;
```

Comment:
I rather we didn't burden `PublicKey` with this.

Rather create a new trait called `DiffieHellmanSharedSecret`

Reply:
Remove ```shared_secret``` from ```PublicKey``` and added a new trait ```DiffieHellmanSharedSecret```